### PR TITLE
Jakarta RESTful Web Services 2.1

### PR DESCRIPTION
**Pull request by spericas**
Signed-off-by: Santiago Pericas-Geertsen <santiago.pericasgeertsen@oracle.com>

## Specification PR template
Include the following:
- [X] A directory in the form wombat/x.y containing the following.
- [X] Specification PDF in the form of wombat_x.y.pdf
- [X] Specification HTML in the form of wombat_x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [ ] For a Release Review, a summary that a Compatible Implementation
      is complete, passes the TCK, and that the TCK includes sufficient
      coverage of the specification. The TCK users guide MUST include
      the instructions to run the compatible implementations used to
      validate the release.  Instructions MAY be by reference.
- [x] The URL of the OSSRH staging repository for the api, javadoc:
    https://oss.sonatype.org/content/repositories/staging/jakarta/ws/rs/jakarta.ws.rs-api/2.1.6/
- [x] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      http://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-restful-ws-tck-2.1.0.zip
- [x] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/jaxrs-api/issues/789

Include the following in PR#2:
- [x] Specification JavaDoc in the wombat/x.y/apidocs directory.

### CODE REVIEW

**Comment by spericas** [comment]
> In the spec document the version should be "2.1"; we don't use dot-dot versions for specs.
> And it should not be marked "DRAFT".

Roger on dot-dot. What should the status be set at if not "DRAFT"?

**Comment by bshannon** [comment]
>   What should the status be set at if not "DRAFT"?

Well, I see that you've discovered a  bug in the template.  :-(
I've been setting it to the empty string in the pom.xml so that nothing is added
to the pdf header/footer, but that doesn't work for the Status: entry at the top
of the document.

Someone with more AsciiDoc skill could probably figure out how to render
that one line as something like "Status: Final Release" if the string is empty.

**Comment by kwsutter** [comment]
There is a \<status\> element in the spec-template pom.xml.  If that is blanked out, then nothing is added to the header or footer.  The "Status:" in the block diagram at the top of the EFSL is blank.  (As Bill pointed out...)

An easy to override this is to pass in the desired string on the mvn command:
```
mvn clean install -Dstatus=Final
```
Depending on your Jenkins jobs for doing a release, this parameter could easily be passed in when doing your "final" build.  Does that help?

**Comment by spericas** [comment]
If the goal is to remove the status line altogether:

```
ifeval::["{revremark}" != ""]
Status: {revremark}
endif::[]
```

**Comment by bshannon** [comment]
> If the goal is to remove the status line altogether:
> 
> ```
> ....
> Specification: {doctitle}
> 
> Version: {revnumber}
> 
> ifeval::["{revremark}" != ""]
> Status: {revremark}
> {nbsp}
> endif::[]
> Release: {revdate}
> ....
> ```

Yes, that's better.  An alternative would be to add an "else" that adds "Status: Final Release".
I just didn't want the status information on every page which is why I blanked it out in the pom before release.

**Comment by spericas** [comment]
> > If the goal is to remove the status line altogether:
> > ```
> > ....
> > Specification: {doctitle}
> > 
> > Version: {revnumber}
> > 
> > ifeval::["{revremark}" != ""]
> > Status: {revremark}
> > {nbsp}
> > endif::[]
> > Release: {revdate}
> > ....
> > ```
> 
> Yes, that's better. An alternative would be to add an "else" that adds "Status: Final Release".
> I just didn't want the status information on every page which is why I blanked it out in the pom before release.

 I'm not aware of such an "advance" construct being available, so:

....
Specification: {doctitle}

Version: {revnumber}

ifeval::["{revremark}" != ""]
Status: {revremark}
endif::[]
ifeval::["{revremark}" == ""]
Status: Final Release
endif::[]

Release: {revdate}
....

I can make this change in my local files and update this PR if that's acceptable.

**Comment by bshannon** [comment]
That looks great to me, thanks!

Now we just need to figure out how to propagate that to all the specs...

File: restful-web-services/2.1/_index.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+date: 2019-07-12T14:29:05-0700
+summary: "First release for Jakarta EE 8"
+---
+Jakarta RESTful Web Services provides a foundational API to develop web services 
+following the Representational State Transfer (REST) architectural pattern.
+
+* [Jakarta RESTful Web Services 2.1 Specification Document](restful-web-services_2.1.pdf) (PDF)
+* [Jakarta RESTful Web Services 2.1 Specification Document](restful-web-services_2.1.html) (HTML)
+* [Jakarta RESTful Web Services 2.1 Javadoc](./apidocs)
+* [Jakarta RESTful Web Services 2.1 TCK](https://jenkins.eclipse.org/jaxrs/job/TCK_build/8/artifact/CTS/release/JAXRS_BUILD/latest/jaxrstck-2.1.0_16-Jul-2019.zip)
```
Comment (bshannon) [inline restful-web-services/2.1/_index.md]: This needs to be the URL to the approved and signed EFTL download, which will be:
https://download.eclipse.org/jakartaee/restful-ws/2.1/eclipse-restful-ws-tck-2.1.0.zip

File: restful-web-services/2.1/_index.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+|Payara                 |         |         |          |
+|Red Hat                |         |         |          |
+|Tomitribe              |         |         |          |
+|EE4J PMC               |         |         |          |
+|Participant Members    |         |         |          |
+|Committer Members      |         |         |          |
+
+# Compatible Implementations
+
+* Eclipse Jakarta RESTful Web Services 2.1.6
```
Comment (bshannon) [inline restful-web-services/2.1/_index.md]: This needs to be the name and a link to a complete Compatible Implementation,
which presumably would be some version of Jersey.

**Comment by jansupol** [comment]
>  template [here](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)

@bshannon Can you please give an insight on the following?
> By checking this box I acknowledge that the Organization I represent accepts the terms of the EFTL.
> By checking this box I attest that all TCK requirements have been met, including any compatibility rules.

**Comment by bshannon** [comment]
> > template [here](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
> 
> @bshannon Can you please give an insight on the following?
> 
> > By checking this box I acknowledge that the Organization I represent accepts the terms of the EFTL.
> > By checking this box I attest that all TCK requirements have been met, including any compatibility rules.

The Organization is the Eclipse Foundation, who has given us permission to check those boxes.

File: restful-ws/_index.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+title: "Jakarta RESTful Web Services Specification 2.1"
+summary: "First release for Jakarta EE 8"
```
Comment (bshannon) [inline restful-ws/_index.md]: The title should not include the version number and the summary should be a copy of the scope statement.

**Comment by NottyCode** [comment]
# Spec Review Checklist

1. Spec PR
  - [x] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [x] Directory of form {spec}/x.y
  - [ ] PDF of form {spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form {spec}-spec-x.y.html ("-spec" preferred but not required)
  - [x] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [x] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [x] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://oss.sonatype.org/conte/repositories/staging/jakarta/{spec}/jakarta-{spec}-api/x.y.z/
  - [x] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [x] Second PR for just apidocs

2. _index.md
  - [x] Link to spec pdf
  - [x] Link to spec html
  - [x] Link to apidocs
  - [x] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip
  - [x] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [x] Name of and link to at least one Compatible Implementation

3. javadocs
  - [x] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [x] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [x] Correct spec title
  - [x] Version number of the form x.y, not x.y.z
  - [x] Correct Eclipse copyright line
  - [x] No DRAFT or SNAPSHOT
  - [x] Correct Logo

5. Spec HTML
  - [x] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [x] EFTL license file, preferably named LICENSE.md
  - [x] User's Guide (or equivalent documentation)
  - [x] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [x] Software requirements listed
  - [x] Installation and configuration described
  - [x] How to run tests
  - [x] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request

**Comment by NottyCode** [comment]
The following issues need resolving:

1. The checklist calls for the pdf and html to be of the form {spec}-spec-x.y.{extension}, but the checked in files use the format {spec}-spec_x.y.{extension}
1. There is no staging repository link to the artifacts
1. There is no link to a compatible certification request, this means section 8 & 9 can't be filled out at this time.
1. The javadocs do not container the ESFL license
1. I can't cross check the javadoc jar with the javadoc PR because there is no staging link for artifacts

**Comment by spericas** [comment]
> The following issues need resolving:
> 
> 1. The checklist calls for the pdf and html to be of the form {spec}-spec-x.y.{extension}, but the checked in files use the format {spec}-spec_x.y.{extension}

There was an inconsistency in the instructions, an underscore is used here:

https://raw.githubusercontent.com/jakartaee/specification-committee/master/spec_page_template.md

I've already changed this once. I can use a dash again if that is final.

> 2. There is no staging repository link to the artifacts

 TCKs are being run:

https://oss.sonatype.org/content/groups/staging/jakarta/ws/rs/jakarta.ws.rs-api/2.1.6/

> 3. There is no link to a compatible certification request, this means section 8 & 9 can't be filled out at this time. I found [this](https://github.com/eclipse-ee4j/jaxrs-api/issues/789) which might be it, but the shahash doesn't match the TCK download.
> 4. The javadocs do not container the ESFL license

 They do contain the license and it is linked in the footers and the main page, and the relevant PR for that is already approved https://github.com/jakartaee/specifications/pull/7

> 5. I can't cross check the javadoc jar with the javadoc PR because there is no staging link for artifacts

 Staging link is above and will be shared with the TCK results.

**Comment by spericas** [comment]
Latest updates:

- Spec files renamed to use a dash instead of an underscore to separate version numbers

- Staging repository for artifacts:

https://oss.sonatype.org/content/groups/staging/jakarta/ws/rs/jakarta.ws.rs-api/2.1.6/

- Compatibility certification request:

https://github.com/eclipse-ee4j/jersey/blob/gh-pages/tck/jakarta-restful-web-services-2.1.md

**Comment by NottyCode** [comment]
>    The following issues need resolving:
>    The checklist calls for the pdf and html to be of the form {spec}-spec-x.y.{extension}, but the checked in files use the format {spec}-spec_x.y.{extension}
> There was an inconsistency in the instructions, an underscore is used here:

I've done a PR to make them consistent: https://github.com/jakartaee/specification-committee/pull/9

>         The javadocs do not container the ESFL license
> They do contain the license and it is linked in the footers and the main page, and the relevant PR for that is already approved #7

I see this now. I have marked it as done.

Can you update the PR description to add the staging url and mark the checkmark done? Also I was expecting a link to a certification issue like this: https://github.com/eclipse-ee4j/jca-api/issues/57 The .md file you linked to looks fine, but we need the issue in the PR description.

I've updated the checklist based on the updates.

**Comment by starksm64** [comment]
The certification request given https://github.com/eclipse-ee4j/jersey/blob/gh-pages/tck/jakarta-restful-web-services-2.1.md is not an issue in the spec project. It is just page somewhere that I can't find other than the repo.

**Comment by bshannon** [comment]
> The certification request given https://github.com/eclipse-ee4j/jersey/blob/gh-pages/tck/jakarta-restful-web-services-2.1.md is not an issue in the spec project. It is just page somewhere that I can't find other than the repo.

Right, the certification request needs to be filed as an issue in the Jakarta RESTful Web Services spec project, as described in the [process document](https://github.com/jakartaee/specification-committee/blob/master/process.adoc#filing-a-certification-request).

**Comment by spericas** [comment]
> > The certification request given https://github.com/eclipse-ee4j/jersey/blob/gh-pages/tck/jakarta-restful-web-services-2.1.md is not an issue in the spec project. It is just page somewhere that I can't find other than the repo.
> 
> Right, the certification request needs to be filed as an issue in the Jakarta RESTful Web Services spec project, as described in the [process document](https://github.com/jakartaee/specification-committee/blob/master/process.adoc#filing-a-certification-request).

 I forgot to include the link to the issue:

https://github.com/eclipse-ee4j/jaxrs-api/issues/789

**Comment by NottyCode** [comment]
@spericas can you edit the PR description to add the links there rather than in a comment please?

**Comment by NottyCode** [comment]
Based on rereview the following issues need to be resolved:

1. The [first PR](https://github.com/jakartaee/specifications/pull/7#issue-298179647) comment needs to be updated to contain the [compliance issue](https://github.com/eclipse-ee4j/jaxrs-api/issues/789).
1. The compatibility request links to the staged version of the TCK and it should be the final location
1. The TCK results summary should be hosted on an eclipse project page, not a link to a github .md file
1. The TCK results summary does not contain all the information in the compliance issue. For the other specifications I've reviewed the TCK results summary includes all the info in the compliance issue. See [transactions](https://eclipse-ee4j.github.io/glassfish/certifications/jakarta-transactions/1.3/TCK-Results) for an example.

**Comment by bshannon** [comment]
Looks good to go now, removing draft.

**Comment by jansupol** [comment]
@NottyCode Should be fixed

**Comment by kwsutter** [comment]
* On ballot completion
 - [x] The designated specification committee representative adds this final checklist to the main PR.
 - [ ] The specification committee representative merges the specification PRs.
 - [ ] The designated specification committee representative calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job.
 - [ ] The specification project member who created the staging release promotes the specification api jars to maven central.
 - [ ] The EMO updates the specification page with the ballot results.
This list goes on the committed spec index page.
 - [ ] The specification project team should go through the merged spec website page to verify all the links are valid.
 - [ ] The specification project team should approve the compatibility request and send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] The specification project team should merge any final release branch as approriate for the branch management for the project.

**Comment by spericas** [comment]
@kwsutter I'm assuming all the tasks in the latest checklist are independent and can be completed in any order. If not, please clarify. Thx.

**Comment by kwsutter** [comment]
> @kwsutter I'm assuming all the tasks in the latest checklist are independent and can be completed in any order. If not, please clarify. Thx.

Yeah, for the most part, they can be independent.  Just a mad rush to the finish line...  :-)

**Comment by spericas** [comment]
- [X] The specification project member who created the staging release promotes the specification api jars to maven central.
- [X] The specification project team should approve the compatibility request and send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.

**Comment by bshannon** [comment]
@spericas You should be able to check off the items in the original [checklist above](https://github.com/jakartaee/specifications/pull/7#issuecomment-527186204).
If not, let me know.  I checked off the two you added above.

**Comment by spericas** [comment]
> @spericas You should be able to check off the items in the original [checklist above](https://github.com/jakartaee/specifications/pull/7#issuecomment-527186204).
> If not, let me know. I checked off the two you added above.

The one @kwsutter added, I cannot update.

**Comment by bshannon** [comment]
Interesting.  Ok, we're learning more about how GitHub permissions work.
We may need to modify our procedures...

**Comment by spericas** [comment]
- [X] The specification project team should go through the merged spec website page to verify all the links are valid.
- [X] The specification project team should merge any final release branch as appropriate for the branch management for the project.

**Pull request merged**